Core 9736/sr allow sd refs

IoannisRP opened the pull request:
Implements: [CORE-9736](https://redpandadata.atlassian.net/browse/CORE-9736)

Relax reference lookup for avro, proto and json to allow for references to be soft deleted.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes

* none


[CORE-9736]: https://redpandadata.atlassian.net/browse/CORE-9736?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pandaproxy/sr: Allow soft-deleted references for proto

IoannisRP committed:
pandaproxy/sr: Allow soft-deleted references for json

IoannisRP committed:
pandaproxy/sr: Allow soft-deleted references for avro

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--541beba3-8364-4d25-94d1-edcb848f07fa-->
<!--TEST RESULTS-->
<!--f273e3f2-49d3-4851-9472-b072c6da76e4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64196">build#64196</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| pandaproxy_schema_registry_single_thread_rpunit.pandaproxy_schema_registry_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64196#01960136-99a9-4457-b097-6574990a8d0a | FAIL | 0/2 |
| pandaproxy_schema_registry_single_thread_rpunit.pandaproxy_schema_registry_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64196#01960136-99aa-4b7b-9fc2-f8d19bd86688 | FAIL | 0/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64196#01960190-9b44-4e65-893e-82398a7bece5 | FLAKY | 1/2 |
| rptest.tests.follower_fetching_test.FollowerFetchingTest.test_with_leadership_transfers | ducktape | https://buildkite.com/redpanda/redpanda/builds/64196#01960190-9b46-4124-8f65-df4d30317638 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64196#0196017d-009b-43f7-8747-ea2dfe708269 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64228">build#64228</a></summary>


| test_id                                                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_record_multiplexer_rpfixture.gtest_record_multiplexer_rpfixture                                                                          | unit        | https://buildkite.com/redpanda/redpanda/builds/64228#01960fba-63d3-4ec0-81a3-d534927af6f5 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64228#01961001-0a14-4cdf-b735-012ee9b873bb | FLAKY         | 1/2      |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64228#01961001-0a13-495e-99aa-e9b7f0d64370 | FLAKY         | 1/2      |
</details>
<!--/body-->

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/7ba86f29d95ebb86f007a27ba3050bd8b49b0519..6c9d3a3f65392fcf3c5cce1e16e99b51f8f14cb1):
- Remove hanging future (leftover from local testing)

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -232,14 +232,12 @@ std::string_view as_string_view(const json::Value& v) {
+          .handle_exception([&](const std::exception_ptr&) -> schema_id {
+              throw as_exception(
+                no_reference_found_for(schema, ref.sub, ref.version));
+          })
+          .then([&](schema_id) { return; });
```
Maybe some more debugging left over?
```suggestion
          });

BenPope commented on src/v/pandaproxy/schema_registry/avro.cc:
```diff
@@ -563,7 +563,7 @@ ss::future<collected_schema> collect_schema(
         if (!collected.contains(ref.name)) {
             auto ss = co_await store.get_subject_schema(
-              ref.sub, ref.version, include_deleted::no);
+              ref.sub, ref.version, include_deleted::yes);
```
Not really related to this PR, but the error that's returned isn't consistent with json, `no_reference_found_for`.

IoannisRP replied:
let's address these separately.

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -438,7 +438,7 @@ ss::future<pb::FileDescriptorProto> build_file_with_refs(
         }
         auto dep = co_await store.get_subject_schema(
-          ref.sub, ref.version, include_deleted::no);
+          ref.sub, ref.version, include_deleted::yes);
```
This is also inconsistent with json, as we don't throw a `no_reference_found_for`

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -232,14 +232,12 @@ std::string_view as_string_view(const json::Value& v) {
-              }
-          });
+        co_await store.get_id(ref.sub, ref.version)
+          .handle_exception([&](const std::exception_ptr&) -> schema_id {
+              throw as_exception(
+                no_reference_found_for(schema, ref.sub, ref.version));
+          })
+          .then([&](schema_id) { return; });
```
I wonder if we should be a little stricter here?
```suggestion
          .handle_exception_type([&](const exception& e) -> schema_id {
              if (
                e.code() == error_code::subject_not_found
                || e.code() == error_code::schema_id_not_found) {
                  throw as_exception(
                    no_reference_found_for(schema, ref.sub, ref.version));
              }
              throw;
          });
```

IoannisRP replied:
i don't think in its current form it could throw anything else, but future-proofing it sounds good. I'll add it :+1:

BenPope reviewed (commented):
LGTM

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/6c9d3a3f65392fcf3c5cce1e16e99b51f8f14cb1..06dbbc6eef7390e03d30f9c19db25c940c8f0d5e):
- addressed review comments

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

dotnwat commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -232,13 +232,15 @@ std::string_view as_string_view(const json::Value& v) {
                   throw as_exception(
                     no_reference_found_for(schema, ref.sub, ref.version));
               }
+              throw;
```
i think you want something like `return ss::make_exception_future` here instead of `throw`. throwing would be the correct thing to do in a coroutine, but this is running in a continuation. it'll be captured properly by `ss::futurize`, but it's not idiomatic.

IoannisRP replied:
Fair point. I'll make a patch for it
